functions/venue.js

```diff
@@ -826,6 +826,17 @@ exports.getOwnerData = functions.https.onCall(async ({ userId }) => {
   return user;
 });
 
+exports.setVenueLiveStatus = functions.https.onCall(async (data, context) => {
+  checkAuth(context);
+
+  const update = {
+    isLive: data.isLive,
+  };
+
+  await admin.firestore().collection("venues").doc(data.venueId).update(update);
```

Comment:
Good idea. Do you have an example of how it should be done?

Reply:
Umm... regex?! :p

Not sure tbh, I don't think we currently do this anywhere, at least not that I have seen.

I would be looking at our 'mangle a `venueName` to make it into a `venueId`' code + the firebase `id` constraints:

- https://firebase.google.com/docs/firestore/quotas#collections_documents_and_fields
  - ![image](https://user-images.githubusercontent.com/753891/116647106-1e779a80-a9bd-11eb-9d77-11ff71284781.png)
  - ![image](https://user-images.githubusercontent.com/753891/116647118-26cfd580-a9bd-11eb-842b-5c2b72479f4a.png)

Reply:
This should be extracted into a function (or maybe a couple) in a common place so that we can reuse them across the rest of our functions here.

I'm not sure if we have it on the backend yet, but it would also be really useful for us to start making use of `Yup` (the same thing we use for frontend validations) on the backend so that we're not blindly trusting the provided data.

Reply:
Is this something you mean?
https://regex101.com/r/ECPGfx/1

Reply:
Ummm.. seems super complex: `^(?!\.\.?$)(?!.*__.*__)([^\/]{1,1500})$`

We don't need to solve every part of this in regex, we can use JS for what is easy/makes sense (eg. the length)

We also don't need to literally follow the firebase restrictions so long as they fit within our own restrictions (and our own restrictions fit the firebase restrictions)

Eg. I think we tend to only user lowercase alpha + numeric in our `venueId`s? Which is currently seemingly handled in a few disparate places across the codebase:

- https://github.com/sparkletown/sparkle/blob/staging/src/api/admin.ts#L153-L154
  - `createUrlSafeName`: `name.replace(/\W/g, "").toLowerCase()`
- https://github.com/sparkletown/sparkle/blob/staging/functions/venue.js#L191-L193
  - `getVenueId`: `name.replace(/\W/g, "").toLowerCase()`
- https://github.com/sparkletown/sparkle/blob/staging/scripts/create-jazz-bar.ts#L59
  - `const venueId = name.replace(/\W/g, "").toLowerCase()`

Reply:
Ummm.. that seems super complex: `^(?!\.\.?$)(?!.*__.*__)([^\/]{1,1500})$`

---

We don't need to solve every part of this in regex, we can use JS for what is easy/makes sense (eg. the length)

We also don't need to literally follow the firebase restrictions so long as they fit within our own restrictions (and our own restrictions fit the firebase restrictions)

Eg. I think we tend to only user lowercase alpha + numeric in our `venueId`s? Which is currently seemingly handled in a few disparate places across the codebase:

- https://github.com/sparkletown/sparkle/blob/staging/src/api/admin.ts#L153-L154
  - `createUrlSafeName`: `name.replace(/\W/g, "").toLowerCase()`
- https://github.com/sparkletown/sparkle/blob/staging/functions/venue.js#L191-L193
  - `getVenueId`: `name.replace(/\W/g, "").toLowerCase()`
- https://github.com/sparkletown/sparkle/blob/staging/scripts/create-jazz-bar.ts#L59
  - `const venueId = name.replace(/\W/g, "").toLowerCase()`

This equates to removing any non-word character from the string (which is then lowercased in the JS):
![image](https://user-images.githubusercontent.com/753891/116839764-deabef80-ac16-11eb-96f9-3c62fca4eb03.png)

Note that this also allows `_` (which I'm not sure if we currently use anywhere.. but we probably don't want to?), so we'll need to account for that in our check as well.

So.. my guess would be that a simplified `venueId` check for us would be that it contains lowercase alphanumeric + `_`, and is below a certain length that we deem appropriate. Perhaps something like:

```
[a-z0-9_]{1,1500}
```

Reply:
Test comment?

Reply:
Done in https://github.com/sparkletown/sparkle/pull/1299/commits/dcab7c233b941a6e482d819adc1faf72ffe801ab